lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,625 @@
+                  requireResidentKey: boolean;
+                  residentKey: "required" | "preferred" | "discouraged";
+                  userVerification: "required" | "preferred" | "discouraged";
+              };
+          }
+        | { status: "EMAIL_MISSING_ERROR" } // email is required if not using session or recoverAccountToken
+        | { status: "SESSION_MISSING_ERROR" } // session is required if not using email or recoverAccountToken
+        | { status: "RECOVER_ACCOUNT_TOKEN_MISSING_ERROR" } // recoverAccountToken is required if not using email or session
+        | { status: "RECOVER_ACCOUNT_TOKEN_INVALID_ERROR" }
+        | { status: "RELYING_PARTY_ID_MISSING_ERROR" }
+        | { status: "RELYING_PARTY_ID_INVALID_ERROR" } // wrong format; must be domain
+        | { status: "RELYING_PARTY_NAME_MISSING_ERROR" } // relyingPartyName is required
+        | { status: "ORIGIN_MISSING_ERROR" }
+        | { status: "ORIGIN_INVALID_ERROR" } // wrong format; must be domain url
+    >;
+
+    signInOptions(input: {
+        relyingPartyId: string;
+        origin: string;
+        userVerification: "required" | "preferred" | "discouraged" | undefined; // see register options
+        timeout: number | undefined;
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<
+        | {
+              status: "OK";
+              webauthnGeneratedOptionsId: string;
+              challenge: string;
+              timeout: number;
+              userVerification: "required" | "preferred" | "discouraged";
+          }
+        | { status: "RELYING_PARTY_ID_MISSING_ERROR" }
+        | { status: "RELYING_PARTY_ID_INVALID_ERROR" } // wrong format; must be domain
+        | { status: "ORIGIN_MISSING_ERROR" }
+        | { status: "ORIGIN_INVALID_ERROR" } // wrong format; must be domain url
+    >;
+
+    signUp(input: {
+        webauthnGeneratedOptionsId: string;
+        credential: {
+            id: string;
+            rawId: string;
+            response: {
+                clientDataJSON: string;
+                attestationObject: string;
+                transports?: ("ble" | "cable" | "hybrid" | "internal" | "nfc" | "smart-card" | "usb")[];
+                userHandle: string;
+            };
+            authenticatorAttachment: "platform" | "cross-platform";
+            clientExtensionResults: Record<string, unknown>;
+            type: "public-key";
+        };
+        session: SessionContainerInterface | undefined;
+        shouldTryLinkingWithSessionUser: boolean | undefined;
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<
+        | {
+              status: "OK";
+              user: User;
+              recipeUserId: RecipeUserId;
+          }
+        | { status: "EMAIL_ALREADY_EXISTS_ERROR" }
+        | { status: "WRONG_CREDENTIALS_ERROR" }
+        | { status: "CREDENTIAL_MISSING_ERROR" }
+        | { status: "GENERATED_OPTIONS_ID_MISSING_ERROR" }
```

Comment:
Whats the diff between GENERATED_OPTIONS_ID_MISSING_ERROR and CREDENTIAL_MISSING_ERROR?

Reply:
done. those are exceptions